server/auth.go

```diff
@@ -344,6 +359,25 @@ func (s *Server) isClientAuthorized(c *client) bool {
 	return true
 }
 
+// returns false if the client needs to be disconnected
+func (s *Server) matchesPinnedCert(c *client, tlsPinnedCerts PinnedCertSet) bool {
+	if tlsPinnedCerts == nil {
+		return true
+	}
+	tlsState := c.GetTLSConnectionState()
```

Comment:
I wonder if this is correct. Basically you assume that because there is a tlsPinnedCerts map provided (for any of the accept ports that we have, clients, routes, etc..), then the connection has to be TLS, and if not, then it failed this test.
I believe that we have a feature that accepts TLS or non TLS on the same port (see option AllowNonTLS, which applies only to clients at the moment, but that includes MQTT). So if we are here, and this is a CLIENT, and tlsState  is nil, this is probably because of AllowNonTLS, so it should be accepted?

Reply:
I mean if you specify a map, then you should fail if no certificate is to be had.

Reply:
I guess you right. If the user configures this, then it is likely that they want to force the use of a certificate.